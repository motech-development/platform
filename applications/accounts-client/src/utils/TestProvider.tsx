import { AuthContext, AuthProvider } from '@motech-development/auth';
import i18n from 'i18next';
import React, { FC, ReactElement } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import { Router } from 'react-router-dom';
import history from '../history';

const user = {
  name: 'Mo Gusbi',
};

export const getIdTokenClaims = jest.fn();

export const getTokenSilently = jest.fn();

export const loginWithRedirect = jest.fn();

export const logout = jest.fn();

export interface ITestProviderProps {
  children: ReactElement;
  isAuthenticated?: boolean;
  isLoading?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  translations?: any;
}

const TestProvider: FC<ITestProviderProps> = ({
  children,
  isAuthenticated = true,
  isLoading = false,
  translations = {},
}) => {
  const testI18n = i18n;

  testI18n.use(initReactI18next).init({
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false,
    },
    lng: 'en',
    resources: {
      en: {
        ...translations,
      },
    },
  });

  return (
    <Router history={history}>
      <AuthProvider>
        <AuthContext.Provider
          value={{
            getIdTokenClaims,
            getTokenSilently,
            isAuthenticated,
            isLoading,
            loginWithRedirect,
            logout,
            user,
          }}
        >
          <I18nextProvider i18n={testI18n}>{children}</I18nextProvider>
        </AuthContext.Provider>
      </AuthProvider>
    </Router>
  );
};

export default TestProvider;
