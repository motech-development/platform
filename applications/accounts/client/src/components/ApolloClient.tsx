import { Apollo } from '@motech-development/appsync-apollo';
import { useAuth } from '@motech-development/auth';
import { Loader } from '@motech-development/breeze-ui';
import React, { FC, memo, ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import Container from './Container';
import ErrorCard from './ErrorCard';

export interface IApolloClientProps {
  children: ReactNode;
}

const ApolloClient: FC<IApolloClientProps> = ({ children }) => {
  const { getTokenSilently, isAuthenticated, isLoading } = useAuth();
  const { t } = useTranslation('apollo-client');

  return (
    <Apollo
      error={
        <Container>
          <ErrorCard
            title={t('error.title')}
            description={t('error.description')}
          />
        </Container>
      }
      fallback={<Loader />}
      getTokenSilently={getTokenSilently}
      isAuthenticated={isAuthenticated}
      isLoading={isLoading}
      unauthorised={
        <Container>
          <ErrorCard
            title={t('unauthorised.title')}
            description={t('unauthorised.description')}
          />
        </Container>
      }
    >
      {children}
    </Apollo>
  );
};

export default memo(ApolloClient);
