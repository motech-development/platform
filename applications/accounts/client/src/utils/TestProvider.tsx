import { AuthContext, AuthProvider } from '@motech-development/auth';
import { ToastContext, ToastProvider } from '@motech-development/breeze-ui';
import i18n from 'i18next';
import React, { FC, ReactElement } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import { Route, Router, Switch } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';

export const buildAuthorizeUrl = jest.fn();

export const getIdTokenClaims = jest.fn();

export const getTokenSilently = jest.fn();

export const loginWithPopup = jest.fn();

export const loginWithRedirect = jest.fn();

export const logout = jest.fn();

export const add = jest.fn(({ onDismiss }) => {
  if (onDismiss) {
    onDismiss();
  }
});

export const remove = jest.fn();

export interface ITestProviderProps {
  children: ReactElement;
  isAuthenticated?: boolean;
  isLoading?: boolean;
  history?: MemoryHistory;
  path?: string;
  user?: object;
}

const TestProvider: FC<ITestProviderProps> = ({
  children,
  isAuthenticated = true,
  isLoading = false,
  path = '/',
  history = createMemoryHistory({
    initialEntries: [path],
  }),
  user = {
    name: 'Mo Gusbi',
    sub: 'user-id',
  },
}) => {
  const testI18n = i18n;

  testI18n.use(initReactI18next).init({
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false,
    },
    lng: 'en',
    resources: {
      en: {},
    },
  });

  return (
    <Router history={history}>
      <AuthProvider>
        <AuthContext.Provider
          value={{
            buildAuthorizeUrl,
            getIdTokenClaims,
            getTokenSilently,
            isAuthenticated,
            isLoading,
            loginWithPopup,
            loginWithRedirect,
            logout,
            user,
          }}
        >
          <ToastProvider>
            <ToastContext.Provider
              value={{
                add,
                remove,
              }}
            >
              <I18nextProvider i18n={testI18n}>
                <Switch>
                  <Route exact path={path} component={() => children} />
                  <Route
                    path="*"
                    component={() => (
                      <div data-testid="next-page">The next page</div>
                    )}
                  />
                </Switch>
              </I18nextProvider>
            </ToastContext.Provider>
          </ToastProvider>
        </AuthContext.Provider>
      </AuthProvider>
    </Router>
  );
};

export default TestProvider;
