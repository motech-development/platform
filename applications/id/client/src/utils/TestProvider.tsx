import { ToastContext, ToastProvider } from '@motech-development/breeze-ui';
import i18n from 'i18next';
import { FC, ReactNode, Suspense } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';

export const add = jest.fn();

export const remove = jest.fn();

export interface ITestProviderProps {
  children: ReactNode;
}

const TestProvider: FC<ITestProviderProps> = ({ children }) => {
  const testI18n = i18n;

  testI18n.use(initReactI18next).init({
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false,
    },
    lng: 'en',
    resources: {
      en: {},
    },
  });

  return (
    <ToastProvider>
      <ToastContext.Provider
        value={{
          add,
          remove,
        }}
      >
        <I18nextProvider i18n={testI18n}>
          <Suspense fallback={<div />}>{children}</Suspense>
        </I18nextProvider>
      </ToastContext.Provider>
    </ToastProvider>
  );
};

export default TestProvider;
