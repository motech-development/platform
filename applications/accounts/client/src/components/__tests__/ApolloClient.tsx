import { render } from '@testing-library/react';
import React from 'react';
import TestProvider from '../../utils/TestProvider';
import ApolloClient from '../ApolloClient';

describe('ApolloClient', () => {
  it('should show error if no AppSync settings are set', async () => {
    const { findByText } = render(
      <TestProvider>
        <ApolloClient>
          <div data-testid="content" />
        </ApolloClient>
      </TestProvider>,
    );

    await expect(findByText('error.title')).resolves.toBeInTheDocument();
  });

  describe('when AppSync settings are set', () => {
    let env: NodeJS.ProcessEnv;

    beforeEach(() => {
      env = {
        ...process.env,
      };

      process.env.REACT_APP_APPSYNC_URL = 'https://appsync.endpoint';
      process.env.REACT_APP_AWS_REGION = 'eu-west-1';
    });

    afterEach(() => {
      process.env = env;
    });

    it('should show spinner if loading', () => {
      const { container } = render(
        <TestProvider isLoading>
          <ApolloClient>
            <div data-testid="content" />
          </ApolloClient>
        </TestProvider>,
      );
      const loader = container.querySelector('circle');

      expect(loader).toBeInTheDocument();
    });

    it('should show error if user is not authenticated', async () => {
      const { findByText } = render(
        <TestProvider isAuthenticated={false}>
          <ApolloClient>
            <div data-testid="content" />
          </ApolloClient>
        </TestProvider>,
      );

      await expect(
        findByText('unauthorised.title'),
      ).resolves.toBeInTheDocument();
    });

    it('should render child', async () => {
      const { findByTestId } = render(
        <TestProvider>
          <ApolloClient>
            <div data-testid="content" />
          </ApolloClient>
        </TestProvider>,
      );

      await expect(findByTestId('content')).resolves.toBeInTheDocument();
    });
  });
});
