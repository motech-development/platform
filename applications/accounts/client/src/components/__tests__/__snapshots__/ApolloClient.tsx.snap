// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApolloClient should contain the correct type policy 1`] = `
{
  "Mutation": {
    "fields": {
      "deleteTransaction": {
        "merge": [Function],
      },
    },
  },
  "Transactions": {
    "fields": {
      "items": {
        "merge": [Function],
        "read": [Function],
      },
    },
    "keyFields": [
      "id",
      "status",
    ],
  },
}
`;
